crates/fuel-core/src/graphql_api/worker_service.rs

```diff
@@ -100,11 +100,18 @@ use futures::{
 use std::{
     borrow::Cow,
     ops::Deref,
+    sync::{
+        atomic::AtomicU32,
+        OnceLock,
+    },
 };
 
 #[cfg(test)]
 mod tests;
 
+// The last known block height that was processed by the GraphQL service.
+pub static LAST_KNOWN_BLOCK_HEIGHT: OnceLock<AtomicU32> = OnceLock::new();
```

Comment:
todo: remove Static and store in Service? Should still be accessible from middleware

Reply:
Is this so performance critical that we can't do a database lookup for this value? I'd default to querying this from the database, to ensure we don't have any weird consistency issues when this field and the db isn't in sync. I guess this isn't a big problem right now, but it just doesn't feel right to allow for inconsistencies in a field used to synchronize with other nodes.

Reply:
We shouldn't use static. Each extension has access to the context and you can get `ReadDatabase` from the context. `ReadDatabase` knows the latest block height.


You can check how `ViewExtension` works

Reply:
I agree that we should fetch the value from the DB (it is probably going to be fetched from the memtable anyway, since it's frequently updated). I don't agree that we should do this with `async_graphql` Extensions, the reason being that support for headers is pretty limited in there. 

1. Headers do not seem to be set when building the query context in `async_graphql` (see https://github.com/async-graphql/async-graphql/blob/7f1791488463d4e9c5adcd543962173e2f6cbd34/src/schema.rs#L927). 
I have confirmed this by applying the patch below to my PR and running the test. The output confirms that the header is not set by the time we execute the query.
```diff
diff --git a/crates/fuel-core/src/schema/balance.rs b/crates/fuel-core/src/schema/balance.rs
index b6b95228e8..956d5961bb 100644
--- a/crates/fuel-core/src/schema/balance.rs
+++ b/crates/fuel-core/src/schema/balance.rs
@@ -60,6 +60,10 @@ impl BalanceQuery {
         #[graphql(desc = "address of the owner")] owner: Address,
         #[graphql(desc = "asset_id of the coin")] asset_id: AssetId,
     ) -> async_graphql::Result<Balance> {
+        println!(
+            "header set: {}",
+            ctx.http_header_contains("REQUIRED_FUEL_BLOCK_HEIGHT")
+        );
         let query = ctx.read_view()?;
         let base_asset_id = *ctx
             .data_unchecked::<ConsensusProvider>()
```
2. Maybe we could get the header somehow from the ExtensionContext, but this is going to be an ad hoc, undocumented solution and likely not readable. 

What I think we should do instead is stick with Axum's middleware: get a handle to the OnchainDatabase from the CombinedDatabase (just because I don't want to wrap the instance to the CombinedDatabase inside an Arc), and use it to fetch the value of the last height in the middleware. 

Let me know what you think.